src/Configuration/src/ConfigServerBase/ConfigServerDiscoveryService.cs

```diff
@@ -19,68 +24,70 @@ internal class ConfigServerDiscoveryService
-        {
-            var discoveryService = ReflectionHelpers.FindType(_discoveryServiceAssemblies, _discoveryServiceTypeNames);
-            MethodInfo method = null;
-
-            if (discoveryService != null)
-            {
-                method = ReflectionHelpers.FindMethod(discoveryService, "GetInstances");
-            }
+            // override register settings so that we register the app twice
+            var cfgBuilder = new ConfigurationBuilder()
+                .AddConfiguration(_configuration)
+                .AddInMemoryCollection(
+                    new Dictionary<string, string>
+                    {
+                        { "Eureka:Client:ShouldRegisterWithEureka", "false" },
+                        { "Consul:Discovery:Register", "false" }
+                    });
 
-            return method;
+            services.AddSingleton<IConfiguration>(cfgBuilder.Build());
+            services.AddDiscoveryClient(_configuration);
+            var startupServiceProvider = services.BuildServiceProvider();
+            _discoveryClient = startupServiceProvider.GetRequiredService<IDiscoveryClient>();
+            _logger.LogDebug("Found Discovery Client of type {DiscoveryClientType}", _discoveryClient.GetType());
         }
 
-        internal IList<IServiceInstance> InvokeGetInstances(MethodInfo method)
+        internal IEnumerable<IServiceInstance> GetConfigServerInstances()
         {
             var attempts = 0;
             var backOff = _settings.RetryInitialInterval;
-            IList<IServiceInstance> instances;
+            IEnumerable<IServiceInstance> instances;
             do
             {
-                instances = null;
                 try
                 {
-                    _logger?.LogDebug("Locating configserver {serviceId} via discovery", _settings.DiscoveryServiceId);
-                    instances = method.Invoke(null, new object[] { _configuration, _settings.DiscoveryServiceId, _logFactory }) as IList<IServiceInstance>;
+                    _logger.LogDebug("Locating configserver {serviceId} via discovery", _settings.DiscoveryServiceId);
+                    instances = _discoveryClient.GetInstances(_settings.DiscoveryServiceId);
                 }
                 catch (Exception e)
                 {
-                    _logger?.LogError(e, "Exception invoking GetInstances() during config server lookup");
+                    _logger.LogError(e, "Exception invoking GetInstances() during config server lookup");
+                    instances = Enumerable.Empty<IServiceInstance>();
                 }
 
-                if (!_settings.RetryEnabled || (instances != null && instances.Count > 0))
+                if (!_settings.RetryEnabled || instances.Any())
                 {
                     break;
                 }
 
                 attempts++;
-                if (attempts < _settings.RetryAttempts)
+                if (attempts <= _settings.RetryAttempts)
```

Comment:
Previously setting `RetryAttempts` to 1 would result in a single total call rather than 1 call with 1 retry